### PR TITLE
EASY-1937 Added swagger API documentation

### PR DESCRIPTION
**Pull request by rkruithof**
Fixes EASY-1937

- [x] process feedback

#### When applied it will
* move the contents of README.md to `docs/01_manual.md`
* add swagger API documentation

#### Where should the reviewer @DANS-KNAW/easy start?
The  `docs/api/api.yml` contents can be checked at https://editor.swagger.io/

### CODE REVIEW

File: docs/api/api.yml [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    The NAAN represents the configured 'name assigning authority number' which for DANS is 73189.
+    It is not neccessary for this service to supply the bag store the bag is stored in.
+
+  license:
+    name: Apache 2.0
+    url: http://www.apache.org/licenses/LICENSE-2.0
+
+servers:
+  - url: 'http://localhost:20160'
+  - url: 'http://test.dans.knaw.nl:20160'
```
Comment (rvanheest) [inline docs/api/api.yml]: this doesn't work anymore. The local VMs do not exist anymore

File: docs/api/api.yml [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        This is the starting place for navigating the Download Service. It returns a description of the service.
+      responses:
+        200:
+          description: Description of the service
+          content:
+            text/plain:
+              schema:
+                type: string
+                description: A plain text, human readable description of the service
+                example: Download Service is running.
```
Comment (rvanheest) [inline docs/api/api.yml]: See for a correct example:

https://github.com/DANS-KNAW/easy-download/blob/0e23f1b11a2c66e48d3e4f8864025cb9480bad36/src/main/scala/nl.knaw.dans.easy.download/EasyDownloadServlet.scala#L40-L43

File: docs/api/api.yml [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        500:
```
Comment (rvanheest) [inline docs/api/api.yml]: when do you get a `500` with this endpoint?

File: docs/api/api.yml [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+          $ref: '#/components/responses/Unavailable'
+
+  /ark:/{NAAN}/{ItemId}:
+    get:
+      summary: Returns an output stream of the file specified by the ItemId.
+      description: |
+        An item is either a directory or file in a bag. The item which is fetched is specified by the item-id, which is the bag-id followed by the local path of the file in the bag.
```
Comment (rvanheest) [inline docs/api/api.yml]: It can only download single files, not directories.
Also it is not just any file in the bag, but only those in the payload of the bag that can be downloaded through here.

File: docs/api/api.yml [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+        example valid ItemId:
+         40594b6d-8378-4260-b96b-13b57beadf7c/path/to/file
+        ```
+
+      parameters:
+        - $ref: '#/components/parameters/ItemId'
+        - $ref: '#/components/parameters/NAAN'
+      responses:
+        200:
+          description: OK, file stream will be returned
```
Comment (rvanheest) [inline docs/api/api.yml]:
```suggestion
          description: The file stream will be returned
```

File: docs/api/api.yml [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+        400:
+          description: Bad Request.
```
Comment (rvanheest) [inline docs/api/api.yml]:
```suggestion
          description: Invalid bag-id; empty or invalid local path in the payload of the bag
```

File: docs/api/api.yml [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+        404:
```
Comment (rvanheest) [inline docs/api/api.yml]: `401` response missing

https://github.com/DANS-KNAW/easy-download/blob/0e23f1b11a2c66e48d3e4f8864025cb9480bad36/src/main/scala/nl.knaw.dans.easy.download/EasyDownloadServlet.scala#L50-L61

File: docs/api/api.yml [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+          description: The item could not be found or may not be visible.
+        500:
```
Comment (rvanheest) [inline docs/api/api.yml]: `408` response missing:

https://github.com/DANS-KNAW/easy-download/blob/0e23f1b11a2c66e48d3e4f8864025cb9480bad36/src/main/scala/nl.knaw.dans.easy.download/EasyDownloadServlet.scala#L72-L84

File: docs/api/api.yml (hunk 7 above)
Comment (rvanheest) [inline docs/api/api.yml]: also `403` response missing

https://github.com/DANS-KNAW/easy-download/blob/0e23f1b11a2c66e48d3e4f8864025cb9480bad36/src/main/scala/nl.knaw.dans.easy.download/EasyDownloadServlet.scala#L72-L84

File: docs/api/api.yml [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+      parameters:
+        - $ref: '#/components/parameters/ItemId'
+        - $ref: '#/components/parameters/NAAN'
+      responses:
+        200:
+          description: OK, file stream will be returned
+        400:
+          description: Bad Request.
+        404:
+          description: The item could not be found or may not be visible.
```
Comment (rvanheest) [inline docs/api/api.yml]: is this about visibility or accessibility?

File: docs/api/api.yml (hunk 1 above)
Comment (rkruithof) [reply docs/api/api.yml]: Deze URL staat nog in meerdere projecten. Hebben we een alternatieve URL die we hier neer kunnen zetten? Ik had hem voor nu zo neergezet als zijnde 'de easy server waarmee je werkt'

File: docs/api/api.yml (hunk 9 above)
Comment (rkruithof) [reply docs/api/api.yml]: This refers to accessibility, which in the comments in the code was written as "in fact: not visible" where visible meant visible to the service. I can make this more specific.

Comment (rkruithof) [reply docs/api/api.yml]: Insufficient accessibility rights are returned with a 403, the code contained a comment saying: 'in fact : not visible' so I chose to include it in this way. I assume visible means visible to the service. I'll revert it to 'could not be found' since that is what it comes down to.

File: docs/api/api.yml (hunk 4 above)
Comment (rkruithof) [reply docs/api/api.yml]: I believe i did specify files earlier in the documentation. This line is just the definition of `Item-Id`, which is used in more projects. I wanted to keep this definition consistent across projects, so i kept it like this. Do you think it makes it unclear? 

(so, an item-id can refer to a bag or directory, but only item-id's of payload files are accepted)

File: docs/api/api.yml (hunk 3 above)
Comment (rkruithof) [reply docs/api/api.yml]: I would imagine this could occur if one of the services this depends on returns an error or a CompositeError occurs in one of the methods invoked by this. Theoretically it SHOULDN'T happen. 

to be honest, choosing to include or not include this is a recurring question I have had, to which there has been no clear answer yet. The question, more specifically is: Do you include general status codes, and to which extent do you include them. There is documentation of other services which already omitted the 408, because it shouldn't occur on a regular basis and feels like a 'general response code', which is why it also wasn't originally included in this documentation. 

In short, the convention of when to include or not to include general status codes is unclear right now.

File: docs/api/api.yml (hunk 1 above)
Comment (rvanheest) [reply docs/api/api.yml]: I'd suggest only keeping the `localhost` variant as '_the easy server you're working with_'

File: docs/api/api.yml (hunk 3 above)
Comment (rvanheest) [reply docs/api/api.yml]: I'd say a nice convention is: "_if it cannot occur, don't put it in_"... Especially with these documentations, where you're basically reverse engineering the documentation from the code, I think you should read the code and see which status can/cannot be returned from a route.

https://github.com/DANS-KNAW/easy-download/blob/a91e45559027d68e9ee8603403daf3fb302dd91a/src/main/scala/nl.knaw.dans.easy.download/EasyDownloadServlet.scala#L40-L43

Given this to be the implementation for `GET /`, I can only conclude the only suitable response code to be `200`.

File: docs/api/api.yml (hunk 4 above)
Comment (rvanheest) [reply docs/api/api.yml]: My reasoning was that you first start talking about an `item` which can be a **directory** or a **file**. Afterwards you state that you can't download a directory (yet). That sounds a bit weird to me.

I see the problem, though... you want to be both precise in your definition of `item` and clarify the restrictions we currently have for this service. Let's keep it as is for now and take another look once we implement downloading a whole bag or directories in a bag.

File: docs/api/api.yml (hunk 1 above)
Comment (janvanmansum) [reply docs/api/api.yml]: You could use `http://localhost:20160` in a development-environment, if you configure the back-end servers to be on `deasy` and/or `ddark`. Other URLs that will be available are:

* https://datavault.dans.knaw.nl/
* https://demo.datavault.dans.knaw.nl/ (maybe).

File: docs/api/api.yml (hunk 4 above)
Comment (janvanmansum) [reply docs/api/api.yml]: * If currently we can only download files, that's what the documentation should say. 
* Just refer to [the definition of file-id](https://dans-knaw.github.io/easy-bag-store/03_definitions.html#item). Your description does not take into account the percent-encoding rules.

File: docs/api/api.yml [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+
+        ```
+        example valid ItemId:
+         40594b6d-8378-4260-b96b-13b57beadf7c/path/to/file
+        ```
+
+      parameters:
+        - $ref: '#/components/parameters/ItemId'
+        - $ref: '#/components/parameters/NAAN'
+      responses:
```
Comment (janvanmansum) [inline docs/api/api.yml]: I would like the description to match the "canonical" description that is defined by the HTTP standard + in cases where it adds essential information: the specifics pertaining to this service. So, for example: the description of 401 could be simply: "401 Unauthorized", because we don't implement any special behavior here.  On the other hand, status 400 may have reasons that are specific to our application.

File: docs/api/api.yml [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+        200:
+          description: The file stream will be returned
+        400:
```
Comment (janvanmansum) [inline docs/api/api.yml]: I wonder when a 400 is returned and when a 404. This has been a source of confusion in the bag-store API itself as well. Does an "invalid bag-id" mean a string that could never be a bag-id, because it is not a UUID, or also a well-formed UUID which does not refer to an existing bag?

File: docs/api/api.yml [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+          description: Invalid bag-id; empty or invalid local path in the payload of the bag.
+        401:
+          description: Invalid credentials were provided.
+        403:
+          description: The provided credentials do not have sufficient access rights to download the file.
+        404:
+          description: The item could not be found.
+        408:
```
Comment (janvanmansum) [inline docs/api/api.yml]: According to https://httpstatuses.com/408, this means the client has opened a connection to the server, but has waited too long to send any request body. Does our code explicitly handle this or is it using this status code improperly?

File: docs/api/api.yml (hunk 4 above)
Comment (rkruithof) [reply docs/api/api.yml]: I updated the description to refer to the file-id instead of the item-id.

File: docs/api/api.yml (hunk 11 above)
Comment (rkruithof) [reply docs/api/api.yml]: In theory this should be the distinction yes, 400 should be returned in case of a malformed UUID, while a 404 should be returned in the case of a non-existent valid bag-id. However the uuid parser still seems to give inconsistent results when it comes to this. 

For instance let's assume the example UUID: `12341234-1234-1234-1234-123412341234` is the only bag-id in a bag-store. 
If any other valid UUID is used then the service will return a 404 as expected.
In the case of a UUID `12341234-1234-1234-1234-12341234123` however a 400 response seems more appropriate, but a 404 is still given.
When trying to take this to the extreme, in the case of a uuid: `0-1-2-3-4` the service still returns a 404 status code with the message: "not found: 00000000-0001-0002-0003-000000000004/data/a/deeper/path/some file.txt" which confirms that the parser will consider the uuid valid as long as it consists of the correct number of `-` separated segments.

In the case of the UUID containing more digits per segment than would be expected the service also returns a 404 instead of a 400.
For example the UUID `12341234-1234-1234-1234-1234123412341` would be parsed as `12341234-1234-1234-1234-234123412341` and would be considered valid.

Comment (rkruithof) [reply docs/api/api.yml]: In easy-bag-store `nl.knaw.dans.easy.bagstore/ItemId.scala` implements the method `validateUuidLength` which checks that a UUID contains exactly 36 characters. If it does not it will return an IllegalArgumentException.

File: docs/api/api.yml (hunk 12 above)
Comment (rkruithof) [reply docs/api/api.yml]: As discussed, 408 does not make sense to pass on to the user, as this status code would represent a server-side timeout, which is different to a client-side one. This response will be removed from the documentation.

File: docs/api/api.yml [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+        400:
+          description: Invalid bag-id; empty or invalid local path in the payload of the bag.
+        401:
+          description: Unauthorized.
+        403:
+          description: Forbidden.
+        404:
+          description: The item could not be found at the specified location.
+        500:
+          description: Service is unavailable because of an error.
```
Comment (janvanmansum) [inline docs/api/api.yml]:
```suggestion
          description: Internal server error.
```

File: docs/api/api.yml (hunk 11 above)
Comment (janvanmansum) [reply docs/api/api.yml]: To avoid confusion, I think it is better to limit 400 to an invalid request *body* and to return 404 for any URL that cannot be resolved (possibly because it contains a malformed UUID).

File: docs/api/api.yml [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+        503:
+          description: The service is unavailable.
+
+components:
+  parameters:
+    FileId:
```
Comment (janvanmansum) [inline docs/api/api.yml]:
```suggestion
    ItemId:
```

File: docs/api/api.yml [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+      in: path
+      name: FileId
```
Comment (janvanmansum) [inline docs/api/api.yml]:
```suggestion
      name: ItemId
```

File: docs/api/api.yml [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+      schema:
+        type: string
+      required: true
+      description: The bag-id of the targeted bag followed by the percent-encoded path location of a file relative to the root of the bag.
```
Comment (janvanmansum) [inline docs/api/api.yml]:
```suggestion
      description: [item-id](https://dans-knaw.github.io/easy-bag-store/03_definitions.html#item-id) of the file to retrieve.
```

File: docs/api/api.yml [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+         40594b6d-8378-4260-b96b-13b57beadf7c/path/to/file
+         40594b6d-8378-4260-b96b-13b57beadf7c/path/to/another%20file
+        ```
+
+      parameters:
+        - $ref: '#/components/parameters/FileId'
+        - $ref: '#/components/parameters/NAAN'
+      responses:
+        200:
+          description: The file stream will be returned.
```
Comment (janvanmansum) [inline docs/api/api.yml]:
```suggestion
          description: OK. Successfully returning the requested file.
```

It seems irrelevant to me that the file might be returned as a stream.

File: docs/api/api.yml [hunk 18]
```diff
@@ -0,0 +1,83 @@
...
+          content:
+            text/plain:
+              schema:
+                type: string
+                description: A plain text, human readable description of the service
+                example: EASY Download Service running v1.1.2.
+
+  /ark:/{NAAN}/{FileId}:
+    get:
+      summary: Returns an output stream of the file specified by the file-id.
```
Comment (janvanmansum) [inline docs/api/api.yml]:
```suggestion
      summary: Returns the file specified by the file-id.
```
